### PR TITLE
Feature: Cold Merge

### PR DESCRIPTION
**Pull request by alhino**
Changes proposed in this pull request:
Added new feature page to describe new enhancements we are planning for Cold Merge

I confirm that this pull request was submitted according to the contribution guidelines: @alhino

This pull request needs review by: @mureinik

Added new feature page to describe new enhancements we are planning
for Cold Merge

### CODE REVIEW

File: source/feature/remove-snapshot.html.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+
```
Comment (rollandf) [inline source/feature/remove-snapshot.html.md]: The file should be located in 'source/develop/release-management/features/storage'

File: source/feature/remove-snapshot.html.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+# Remove Snapshot
```
Comment (rollandf) [inline source/feature/remove-snapshot.html.md]: Missing the headers like for example :
title: DiscardAfterDelete
category: feature
authors: ishaby
feature_name: Discard after delete
feature_modules: vdsm
feature_status: Partially implemented in oVirt 4.0

**Comment by rollandf** [comment]
The CI is failing due to the missing headers I mentioned earlier.

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+Use qemu-img commit instead of qemu-img rebase when deleting snapshots while the VM is down.
+
+## Owner
+
+*   Name: Ala Hino
+*   Email: <ahino@redhat.com>
+
+## Terminology
+
+This feature page is about removing snapshots when the VM is down. As today we use 'Live Merge' when removing snapshots while the VM is running, I will use 'Cold Merge' when referring to removing snapshots while the VM is down.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: Who is "we"  and "I"?
Maybe something like:
"As today the 'Live Merge' term is used [.....], 'Cold Merge' will be used when [....]

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+## Background
+
+Initially, when Cold Merge feature implemented, qemu-img didn't support 'commit' operations hence, qemu-img 'rebase' performed. In certian use cases, qemu-img rebase could take long time, especially when the base volume is significantly larger than the top volume. In these use cases, Live Merge was faster.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: "was implemented"
"certain"

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+Now that qemu-img support 'commit' operations, we want to use this support to implement Cold Merge.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: who is "we" ?

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+## Benefit to oVirt
+
+Removing snapshots will work faster also when the VM is not running.
+
+## Design
+
+During the design of this feature, we want to keep in mind following:
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: who is "we'?
"keep in mind the following"

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+* SPM is a bottleneck and we want to reduce SPM tasks as much as possible
+* Live and Cold merge flows are equivalent so it is a good chance to have a single implementatin for both flows. The flow basically contains following steps:
+
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: "implementation"
"contains the following"

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+* MergeExtendCommand
+
+* ColdMergeCommand: a new command that will execute SDM_MERGE new verb. This command extends StorageJobCommand and will use StorageJobCallback to monitor the merge (host) job
+
+* MergeStatusCommand
+
+* DestroyImageCommand
+
+* ReduceImageCommand: a new command that will reducew the image to optimal size
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: "reduce"

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+### Vdsm
+
+Following new verbs are introduced:
+
+* sdm_merge: another data operation that merges data from top volume to base volume
+
+* Volume.getWatermarks: returns the volume watermarks (will be used to reduce volume size to optimal. See next vern)
+
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: "verb"

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  * The job wasn't found on the host. In this case, the engine checks the base value state, if it is LEGAL, the merge didn't start and the engine has to send merge command again; if the state is ILLEGAL, the engine fails the merge operation and the admin has to retry
+
+* Vdsm fails to get Engine command
+
+  Operation fails and the admin needs to try again
+
+* Vdsm fails to return response to Engine
+
+  Usually engine gets timeout once fails to get Vdsm response. In this case, as done in Live Merge, the operation fails and the admin needs to try merge again. However, where in Live Merge reporting errors to engine was naive, in this case Vdsm will report the exact errors to the engine: a job is running, base/top doesn't exist, etc.
+
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: "once it failed"

**Comment by mykaul** [comment]
@alhino - why is not in yet? Anything that needs to be reviewed here, or can I merge it?

**Comment by alhino** [comment]
Not yet. I want to update it to reflect latest changes

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
+---
+title: Remove Snapshot
+category: feature
+authors: ahino
+wiki_category: Feature|RemoveSnapshot
+wiki_title: Features/Remove Snapshot
+wiki_revision_count: 1
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: No need for any "wiki" tags

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+wiki_last_updated: 2016-09-27
+feature_name: Remove Snapshot
+feature_modules: engine/vdsm
+feature_status: Released
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: Released in 4.1

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+---
+
+# Remove Snapshot
+
+## Summary
+
+Use qemu-img commit instead of qemu-img rebase when deleting snapshots while the VM is down.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: `qemu-img commit`
`qemu-img rebase`

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+*   Name: Ala Hino
+*   Email: <ahino@redhat.com>
+
+## Terminology
+
+This feature page is about removing snapshots when the VM is down. Today, the term 'Live Merge' is used to describe removing snapshots while the VM is running. I will use 'Cold Merge' when referring to removing snapshots while the VM is down.
+
+## Background
+
+Initially, when Cold Merge feature was implemented, qemu-img didn't support 'commit' operations hence, qemu-img 'rebase' performed. When running qemu-img rebase, a temporary volume is created, and data is copied from base volume to the temporary volume. In certain, is not in most, use cases, qemu-img rebase could take long time, because the base volume is significantly larger than the top volume.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: `commit`
`rebase`

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+    * Update base allocation
+
+        For now, base volume allocation is updated to worse case, i.e. to minimum of top-size + size and to maximum allocation:
+
+        potential-alloc = base-size + top-size
+        max-alloc = base-capacity * 1.1 (QCOW-OVERHEAD)
+        new-alloc = min(potential-alloc, max-alloc)
+
+        In the future the plan is to use qemu-img map to calculate the accurate allocation size.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: `qemu-image map`

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+* Merge
+
+    Merge step is another data operation command and is implements as an SDM job.
+    In this step, qemu-img commit is used to perform the merge.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: `qemu-img commit`

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+* Finalize
+
+    Prepare step updates Vdsm metadata and must be run as SPM task.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: ? seems copy/paste issue here

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+    * Mark base volume as LEGAL
+
+        At this point, merge successfully completed, Vdsm and qemu chains updated, hence it is safe now to mark the base volume as LEGAL.
+
+* Destroy image
+    Remove the top volume that was merged.
+
+### Recovery
+
+Today, no recovery special handling is required because data is copiedfrom base volume to top. This copy cannot end with data corruption hence, on failures, the admin simply needs to retry. However, when using qemu-img commit, data is copied from top to base, similar to Live Merge. In this case, recovery isn't as trivial and recovery is performed based on the failure step:
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: copiedfrom -> copied from

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+
+* Engine reboot during Cold Merge
+
+  If engine goes down during Cold Merge, there are two options when it comes up again:
+
+  * The job still exists at the storage. In this case, we continue monitoring the job
+
+  * The job doesn't exist at the storage. In this case, as done in Live Merge, we continue to Merge Status and check whether the volume exists in the chain. If not, we continue the flow. Otherwise, we fail the operation
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: we ?

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+
+* No connectivity to host during Cold Merge
+
+  Once connectivity reestablished, the engine tries to continue monitoring the job on the host. There are two options:
+  
+  * The job exists on the host. In this case, the egine continues to monitor it
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: egine -> engine

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 11 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: removed

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 12 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: done

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 13 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: done

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 14 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: done

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 15 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: dobe

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 16 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: done

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 17 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: fixed

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 18 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: fixed

File: source/develop/release-management/features/storage/remove-snapshot.html.md (hunk 19 above)
Comment (alhino) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: fixed

File: Gemfile.lock [hunk 21]
```diff
@@ -240,6 +240,3 @@ DEPENDENCIES
   therubyracer
   wdm (~> 0.1.0)
   wikicloth
-
```
Comment (rollandf) [inline Gemfile.lock]: Should not be in the commit

File: source/develop/release-management/features/storage/remove-snapshot.html.md [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+feature_name: Remove Snapshot
+feature_modules: engine/vdsm
+feature_status: Released in 4.1
+---
+
+# Remove Snapshot
+
+## Summary
+
+Use `qemu-img commit' instead of 'qemu-img rebase' when deleting snapshots while the VM is down.
```
Comment (rollandf) [inline source/develop/release-management/features/storage/remove-snapshot.html.md]: fix the backticks, should be all `

**Pull request merged**